JsonSubTypes/JsonSubtypesConverter.cs

```diff
@@ -0,0 +1,104 @@
+using Newtonsoft.Json.Linq;
+
+namespace JsonSubTypes
+{
+    //  MIT License
+    //  
+    //  Copyright (c) 2017 Emmanuel Counasse
+    //  
+    //  Permission is hereby granted, free of charge, to any person obtaining a copy
+    //  of this software and associated documentation files (the "Software"), to deal
+    //  in the Software without restriction, including without limitation the rights
+    //  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    //  copies of the Software, and to permit persons to whom the Software is
+    //  furnished to do so, subject to the following conditions:
+    //  
+    //  The above copyright notice and this permission notice shall be included in all
+    //  copies or substantial portions of the Software.
+    //  
+    //  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    //  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    //  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    //  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    //  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    //  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    //  SOFTWARE.
+
+    internal class JsonSubtypesConverter : JsonSubtypes
+    {
+        private readonly bool _serializeDiscriminatorProperty;
+        private readonly Dictionary<Type, object> _supportedTypes = new Dictionary<Type, object>();
+        private readonly Type _baseType;
+        private readonly Dictionary<object, Type> _subTypeMapping;
+
+        [ThreadStatic] private static bool _isInsideWrite;
+        [ThreadStatic] private static bool _allowNextWrite;
+
+        internal JsonSubtypesConverter(Type baseType, string discriminatorProperty,
+            Dictionary<object, Type> subTypeMapping, bool serializeDiscriminatorProperty) : base(discriminatorProperty)
+        {
+            _serializeDiscriminatorProperty = serializeDiscriminatorProperty;
+            _baseType = baseType;
+            _subTypeMapping = subTypeMapping;
+            foreach (var type in _subTypeMapping)
+            {
+                _supportedTypes.Add(type.Value, type.Key);
+            }
+        }
+
+        protected override Dictionary<object, Type> GetSubTypeMapping(Type type)
+        {
+            return _subTypeMapping;
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == _baseType || _supportedTypes.ContainsKey(objectType);
+        }
+
+        public override bool CanWrite
+        {
+            get
+            {
+                if (!_serializeDiscriminatorProperty)
+                    return false;
+
+                if (!_isInsideWrite)
+                    return true;
+
+                if (_allowNextWrite)
+                {
+                    _allowNextWrite = false;
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove this set, which updates a 'static' field from an instance property. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)